dot/state/epoch.go

```diff
@@ -365,3 +454,143 @@ func (s *EpochState) SkipVerify(header *types.Header) (bool, error) {
+func (s *EpochState) StoreBABENextConfigData(epoch uint64, hash common.Hash, val types.NextConfigData) {
+	s.nextEpochLock.Lock()
+	defer s.nextEpochLock.Unlock()
+
+	_, has := s.nextConfigData[epoch]
+	if !has {
+		s.nextConfigData[epoch] = make(map[common.Hash]types.NextConfigData)
+	}
+	s.nextConfigData[epoch][hash] = val
+}
+
+// FinalizeBABENextEpochData stores the right types.NextEpochData by
+// getting the set of hashes from the received epoch and for each hash
+// check if the header is in the database then it's been finalized and
+// thus we can also set the corresponding EpochData in the database
+func (s *EpochState) FinalizeBABENextEpochData(epoch uint64) error {
+	s.nextEpochLock.Lock()
+	defer s.nextEpochLock.Unlock()
+
+	finalizedNextEpochData, err := s.lookForPersistedHashForEpochData(epoch)
+	if err != nil {
+		return fmt.Errorf("cannot find next epoch data: %w", err)
+	}
+
+	ed, err := finalizedNextEpochData.ToEpochData()
+	if err != nil {
+		return fmt.Errorf("cannot transform epoch data: %w", err)
+	}
+
+	err = s.SetEpochData(epoch, ed)
+	if err != nil {
+		return fmt.Errorf("cannot set epoch data: %w", err)
+	}
+
+	// remove previous epochs from the memory
+	epochsToDelete := make([]uint64, 0)
+	for e := range s.nextEpochData {
+		if e <= epoch {
+			epochsToDelete = append(epochsToDelete, e)
+		}
+	}
+
+	for _, e := range epochsToDelete {
+		delete(s.nextEpochData, e)
+	}
+
+	return nil
+}
+
+// FinalizeBABENextConfigData stores the right types.NextConfigData by
+// getting the set of hashes from the received epoch and for each hash
+// check if the header is in the database then it's been finalized and
+// thus we can also set the corresponding NextConfigData in the database
+func (s *EpochState) FinalizeBABENextConfigData(epoch uint64) error {
+	s.nextEpochLock.Lock()
+	defer s.nextEpochLock.Unlock()
+
+	finalizedNextConfigData, err := s.lookForPersistedHashForConfigData(epoch)
+	if err != nil {
+		return fmt.Errorf("cannot find next config data: %w", err)
+	}
+
+	cd := finalizedNextConfigData.ToConfigData()
+	err = s.SetConfigData(epoch, cd)
+	if err != nil {
+		return fmt.Errorf("cannot set config data: %w", err)
+	}
+
+	// remove previous epochs from the memory
+	epochsToDelete := make([]uint64, 0)
+	for e := range s.nextConfigData {
+		if e <= epoch {
+			epochsToDelete = append(epochsToDelete, e)
+		}
+	}
+
+	for _, e := range epochsToDelete {
+		delete(s.nextConfigData, e)
+	}
+
+	return nil
+}
+
+func (s *EpochState) lookForPersistedHashForEpochData(epoch uint64) (types.NextEpochData, error) {
+	epochData, has := s.nextEpochData[epoch]
+	if !has {
+		return types.NextEpochData{}, errEpochNotFound
+	}
+
+	for hash, inMemoryEpochData := range epochData {
+		persisted, err := s.blockState.HasHeaderInDatabase(hash)
+		if err != nil {
+			return types.NextEpochData{}, fmt.Errorf("failed to check header exists at database: %w", err)
+		}
+
+		if !persisted {
+			continue
+		}
+
+		return inMemoryEpochData, nil
```

Comment:
I'd add a comment here to explain what's happening, since it's not obvious. same for `lookForPersistedHashForConfigData`